storage/pin/pin_test.go

```diff
@@ -0,0 +1,328 @@
+//         a) Get all the chunks
+//            get it from retrievalDataIndex
+//            since the assumption is the DB has only this file, it gives all the file's chunks.
+//            getAllRefs cannot be used here as it does not give the chunks that belong to manifests.
+//         b) Get all chunks that are pinned (from pinIndex)
+//            In every upload.. an empty manifest is uploaded. that why add this hash to this list
+//         c) Check if both the above lists are equal
+//   3) Check if all the chunks pinned have the proper pinCounter
+//         -  This is just a simple go through of all the pinned chunks list and check if the counter is
+//            equal to the pin counter given as argument
+func failIfNotPinned(t *testing.T, p *PinAPI, rootHash string, pinCounter uint64, isRaw bool) {
+
+	t.Helper()
+
+	// 1 - Check if the root hash is pinned in pinFilesIndex
+	pinnedFiles := p.getPinnedFiles()
+	if _, ok := pinnedFiles[rootHash]; !ok {
+		t.Fatalf("File %s not pinned in pinFilesIndex", rootHash)
+	}
+
+	// 2a - Get all the chunks of the file from retrievalDataIndex (since the file is uploaded in an empty database)
+	chunksInDB := p.getAllChunksFromDB()
+
+	// 2b - Get pinned chunks details from pinning indexes
+	pinnedChunks := p.collectPinnedChunks(rootHash, "")
+	if !isRaw {
+		// Add the empty manifest chunk
+		pinnedChunks["8b634aea26eec353ac0ecbec20c94f44d6f8d11f38d4578a4c207a84c74ef731"] = pinCounter
+	}
+
+	// 2c - Check if number of chunk hashes are same
+	if len(chunksInDB) != len(pinnedChunks) {
+		t.Fatalf("Expected number of chunks to be %d, but is %d", len(chunksInDB), len(pinnedChunks))
+	}
+
+	// 2c - Check if all the chunk address are same
+	// tolerate 1 chunk failure for dummy manifest on encrypted collection
+	tolerateOneChunkFailure := true
+	for hash := range chunksInDB {
+		if _, ok := pinnedChunks[hash]; !ok {
+			if !isRaw && tolerateOneChunkFailure {
+				tolerateOneChunkFailure = false
+				continue
+			}
+			t.Fatalf("Expected chunk %s not present", hash)
+		}
+	}
+
+	// 3 - Check for pin counter correctness
+	if pinCounter != 0 {
+		for _, pc := range pinnedChunks {
+			if pc != pinCounter {
+				t.Fatalf("Expected pin counter %d got %d", pinCounter, pc)
+			}
+		}
+	}
+}
+
+func failIfUnpinned(t *testing.T, p *PinAPI, rootHash string) {
+
+	t.Helper()
+
+	// root hash should not be in pinFilesIndex
+	pinnedFiles := p.getPinnedFiles()
+	if _, ok := pinnedFiles[rootHash]; ok {
+		t.Fatalf("File %s pinned in pinFilesIndex", rootHash)
+	}
+
+	// The chunks of this file should not be in pinIndex too
+	pinnedChunks := p.collectPinnedChunks(rootHash, "")
+	if len(pinnedChunks) != 0 {
+		t.Fatalf("Chunks of this file present in pinIndex")
+	}
+}
+
+func pinUnpinAndFailIfError(t *testing.T, p *PinAPI, rootHash string, noOfPinUnpin int, isRaw bool) {
+
+	t.Helper()
+
+	// Pin the file and check if it is pinned
+	for i := 0; i < noOfPinUnpin; i++ {
```

Comment:
what does `noOfPinUnpin` mean?

Reply:
this is a counter which is used to pin and unpin multiple times.